### PR TITLE
Support Storage Partitioning

### LINKED ISSUE #382: Support Storage Partitioning
Right now IndexedDB is partitioned by the frame's origin, but we should partition third-party iframes independently from first-party frames: https://docs.google.com/document/d/1xd6MXcUhfnZqIe5dt2CTyCn6gEZ7nOezAEWS0W9hwbQ/edit?resourcekey=0-Z5Xn1i5GJ4rit6_93QKIvA

Every time 'origin' is referenced in the spec, we should replace it with a placeholder for origin or double key'd origin depending on how the storage is partitioned.

### PR DESCRIPTION
**Pull request by arichiv**
This PR is more of a find and replace right now, not all the concepts translate perfectly as https://html.spec.whatwg.org/multipage/webappapis.html#environment-settings-objects needs a storage key and https://storage.spec.whatwg.org/#storage-keys needs to define the opaque concept.

Closes #382

The following tasks have been completed:

 * [X] Confirmed there are no ReSpec/BikeShed errors or warnings.
 * [X] [Modified Web platform tests](https://github.com/web-platform-tests/wpt/pull/34187)

Implementation commitment:

 * [ ] WebKit (https://bugs.webkit.org/show_bug.cgi?id=)
 * [X] Chromium (https://bugs.chromium.org/p/chromium/issues/detail?id=1218100)
 * [ ] Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=)

***
<a href="https://pr-preview.s3.amazonaws.com/arichiv/IndexedDB/pull/383.html" title="Last updated on Jun 8, 2022, 5:20 PM UTC (a123380)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/IndexedDB/383/e196c63...arichiv:a123380.html" title="Last updated on Jun 8, 2022, 5:20 PM UTC (a123380)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -2215,17 +2218,17 @@ The <dfn method for=IDBFactory>open(|name|, |version|)</dfn> method steps are:
 
 1. Let |environment| be [=/this=]'s [=/relevant settings object=].
 
-1. Let |origin| be |environment|'s [=environment settings object/origin=].
+1. Let |storageKey| be the result of running [=obtain a storage key=] given |environment|.
 
-1. If |origin| is an [=opaque origin=], [=throw=] a
+1. If |storageKey|'s origin is an opaque, [=throw=] a
```
Comment (inexorabletash) [inline index.bs:2223]: Change "an opaque" to "opaque" ? Also, can we [link the term](https://html.spec.whatwg.org/multipage/origin.html#concept-origin-opaque) while we're here?

File: index.bs [hunk 2]
```diff
@@ -2286,17 +2289,17 @@ The <dfn method for=IDBFactory>deleteDatabase(|name|)</dfn> method steps are:
...
```
Comment (inexorabletash) [inline index.bs:2294]: Same here ("an opaque" &rarr; "opaque", and linkify)

File: index.bs [hunk 3]
```diff
@@ -2342,16 +2345,16 @@ The <dfn method for=IDBFactory>databases()</dfn> method steps are:
 
 1. Let |environment| be [=/this=]'s [=/relevant settings object=].
 
-1. Let |origin| be |environment|'s [=environment settings object/origin=].
+1. Let |storageKey| be the result of running [=obtain a storage key=] given |environment|.
 
-1. If |origin| is an [=opaque origin=],
+1. If |storageKey|'s origin is an opaque,
```
Comment (inexorabletash) [inline index.bs:2350]: ditto

File: index.bs [hunk 4]
```diff
@@ -6604,10 +6607,10 @@ of user tracking:
     User agents may require the user to authorize access to databases
     before a site can use the feature.
 
-: Origin-tracking of stored data
+: Storage Key tracking of stored data
 ::
-    User agents may record the [=/origins=] of sites that contained content
-    from third-party origins that caused data to be stored.
+    User agents may record the [=/storage key=]s of sites that contained content
```
Comment (inexorabletash) [inline index.bs:6612]: You should be able to put the 's' inside the link; bikeshed is smart enough to handle (some) plural forms.

File: index.bs [hunk 5]
```diff
@@ -6604,10 +6607,10 @@ of user tracking:
...
+    from third-party [=/storage key=]s that caused data to be stored.
```
Comment (inexorabletash) [inline index.bs:6613]: Does this section describing "third-party origins" still make sense when storage keys are involved?

I'd leave this section alone (i.e. revert these changes) and perhaps add a paragraph about storage keys partitioning third-party storage? Or leave a TODO somewhere (new issue?) to add it.

File: index.bs [hunk 6]
```diff
@@ -6715,12 +6718,12 @@ user's wish list; or a hostile site could set a user's session
 identifier to a known ID that the hostile site can then use to track
 the user's actions on the victim site.
 
-Thus, strictly following the <span>origin</span> model described in
+Thus, strictly following the <span>storage key</span> model described in
 this specification is important for user security.
 
-If origins or database names are used to construct paths for
+If [=/storage key=]s or database names are used to construct paths for
```
Comment (inexorabletash) [inline index.bs:6724]: Maybe we should replace this with "host names" instead?

File: index.bs [hunk 7]
```diff
@@ -6715,12 +6718,12 @@ user's wish list; or a hostile site could set a user's session
...
 persistence to a file system they must be appropriately escaped to
-prevent an adversary from accessing information from other origins
+prevent an adversary from accessing information from other [=/storage key=]s
```
Comment (inexorabletash) [inline index.bs:6726]: s can go inside the link

File: index.bs [hunk 8]
```diff
@@ -34,6 +34,7 @@ spec: html; urlPrefix: https://html.spec.whatwg.org/multipage/
     urlPrefix: dom.html
         type: interface
             text: Document; url: document
+        type:event; text:close
```
Comment (inexorabletash) [inline index.bs:37]: What's this for?

File: index.bs [hunk 9]
```diff
@@ -65,6 +66,8 @@ spec: ecma262; urlPrefix: https://tc39.github.io/ecma262/
 spec: storage; urlPrefix: https://storage.spec.whatwg.org/
     type: dfn
         text: storage bucket; url: storage-bucket
+        text: storage key; url: storage-key
```
Comment (inexorabletash) [inline index.bs:69]: Do we want to call this out as temporary anywhere? Or is there a plan for these to be exported from storage?

Comment (arichiv) [reply index.bs:69]: I'm not sure what the plan is @annevk

File: index.bs (hunk 8 above)
Comment (arichiv) [reply index.bs:37]: This fixes a compile error I got, not sure why it cropped up now but it couldn't distinguish between this close and one in serviceworker

Comment (inexorabletash) [reply index.bs:37]: Well, that should go into `<pre class=link-defaults>` not `<pre class=anchors>`... but it exists in main, let me see what's up...

Comment (inexorabletash) [reply index.bs:37]: Okay, this is due to linking weirdness elsewhere - it should be a reference internal to the spec. I'll put up a fix.

Comment (inexorabletash) [reply index.bs:37]: https://github.com/w3c/IndexedDB/pull/384

File: index.bs (hunk 9 above)
Comment (inexorabletash) [reply index.bs:69]: Yeah, that's https://github.com/w3c/IndexedDB/pull/334 which got complicated and stalled since as written it _also_ relied on non-exported concepts. I'll need to stare at it for a while to remember how far along that was.

**Comment by inexorabletash** [comment]
FYI that I made an update to https://github.com/w3c/IndexedDB/pull/334 which could use more eyeballs, and would simplify/unblock this.

**Comment by arichiv** [comment]
Could we get this in first? It seems correct as is and #334 has been in-progress for two years.

File: index.bs [hunk 10]
```diff
@@ -6604,10 +6607,10 @@ of user tracking:
     User agents may require the user to authorize access to databases
     before a site can use the feature.
 
-: Origin-tracking of stored data
+: Storage Key tracking of stored data
```
Comment (inexorabletash) [inline index.bs:6610]: This section was confusing before, and I don't think this improves the situation. Maybe rename to "Attribution of third-party storage" ?

**Review by inexorabletash — CHANGES_REQUESTED** [review]
#334 is probably pretty close now, I just hadn't spent any time on it. Just need to replace "pumpkin" with something. We definitely need #384 - hopefully I can get a reviewer to chime in soon.

File: index.bs (hunk 8 above)
[the remark this answers is not in the document]
Comment (inexorabletash) [reply index.bs:37]: #384 is merged, so this can be removed now.

File: index.bs [hunk 11]
```diff
@@ -6717,12 +6712,12 @@ user's wish list; or a hostile site could set a user's session
 identifier to a known ID that the hostile site can then use to track
 the user's actions on the victim site.
 
-Thus, strictly following the <span>origin</span> model described in
+Thus, strictly following the <span>storage key</span> model described in
```
[the remark this answers is not in the document]
Comment (inexorabletash) [reply index.bs:6715]: (I'm not sure why there were spans there to begin with)

**Review by inexorabletash — COMMENTED** [review]
Some last comments/suggestions, then we can probably merge.

**Review by inexorabletash — APPROVED** [review]
